packages/core/echo/echo-pipeline/src/automerge/collection-synchronizer.ts

```diff
@@ -0,0 +1,197 @@
+  queryCollectionState: (collectionId: string, peerId: PeerId) => void;
+  shouldSyncCollection: (collectionId: string, peerId: PeerId) => Promise<boolean>;
+};
+
+/**
+ * Implements collection sync protocol.
+ */
+export class CollectionSynchronizer extends Resource {
+  private readonly _sendCollectionState: CollectionSynchronizerParams['sendCollectionState'];
+  private readonly _queryCollectionState: CollectionSynchronizerParams['queryCollectionState'];
+  private readonly _shouldSyncCollection: CollectionSynchronizerParams['shouldSyncCollection'];
+
+  /**
+   * CollectionId -> State.
+   */
+  private readonly _perCollectionStates = new Map<string, PerCollectionState>();
+
+  private readonly _connectedPeers = new Set<PeerId>();
+
+  public readonly remoteStateUpdated = new Event<{ collectionId: string; peerId: PeerId }>();
+
+  constructor(params: CollectionSynchronizerParams) {
+    super();
+    this._sendCollectionState = params.sendCollectionState;
+    this._queryCollectionState = params.queryCollectionState;
+    this._shouldSyncCollection = params.shouldSyncCollection;
+  }
+
+  protected override async _open(ctx: Context): Promise<void> {
+    scheduleTaskInterval(
+      this._ctx,
+      async () => {
+        for (const collectionId of this._perCollectionStates.keys()) {
+          this.refreshCollection(collectionId);
+        }
+      },
+      POLL_INTERVAL,
+    );
+  }
+
+  getRegisteredCollectionIds(): string[] {
+    return [...this._perCollectionStates.keys()];
+  }
+
+  getLocalCollectionState(collectionId: string): CollectionState | undefined {
+    return this._getPerCollectionState(collectionId).localState;
+  }
+
+  setLocalCollectionState(collectionId: string, state: CollectionState) {
+    this._getPerCollectionState(collectionId).localState = state;
+
+    queueMicrotask(async () => {
+      for (const peerId of this._connectedPeers) {
+        await this._refreshInterestedPeers(collectionId, peerId);
+      }
+
+      this.refreshCollection(collectionId);
+    });
+  }
+
+  getRemoteCollectionStates(collectionId: string): ReadonlyMap<PeerId, CollectionState> {
+    return this._getPerCollectionState(collectionId).remoteStates;
+  }
+
+  refreshCollection(collectionId: string) {
+    const state = this._getPerCollectionState(collectionId);
+    for (const peerId of this._connectedPeers) {
+      if (state.interestedPeers.has(peerId)) {
+        const lastQueried = state.lastQueried.get(peerId) ?? 0;
+        if (Date.now() - lastQueried > MIN_QUERY_INTERVAL) {
+          state.lastQueried.set(peerId, Date.now());
+          this._queryCollectionState(collectionId, peerId);
+        } else {
+          scheduleTask(this._ctx, () => this.refreshCollection(collectionId), MIN_QUERY_INTERVAL);
+        }
+      }
+    }
+  }
+
+  /**
+   * Callback when a connection to a peer is established.
+   */
+  onConnectionOpen(peerId: PeerId) {
+    this._connectedPeers.add(peerId);
+
+    queueMicrotask(async () => {
```

Comment:
why do we refreshCollection for all peers when a new connection is open. maybe make it just
```ts
    queueMicrotask(async () => {
      for (const [collectionId, state] of this._perCollectionStates.entries()) {
        if (await this._shouldSyncCollection(collectionId, peerId)) {
          state.interestedPeers.add(peerId);
          state.lastQueried.set(peerId, Date.now());
          this._queryCollectionState(collectionId, peerId);
        }
      }
    });
```
instead?